cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -1181,4 +1191,30 @@ public void deleteState(AppStateKey request) throws ApplicationNotFoundException
   public void deleteAllStates(NamespaceId namespaceId, String appName) throws ApplicationNotFoundException {
     store.deleteAllStates(namespaceId, appName);
   }
+
+  private void emitMetrics(String namespace, String appName, String metricName) {
+    Map<String, String> tags = ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, namespace,
+                                               Constants.Metrics.Tag.APP, appName);
+    metricsCollectionService.getContext(tags).increment(metricName, 1);
+  }
+
+  private void emitMetricForMultipleSourcePlugins(boolean isPreview,
+                                                  ApplicationWithPrograms applicationWithPrograms)
+    throws ApplicationNotFoundException {
+    if (!isPreview) {
+      int sourceCount = 0;
+      ApplicationId appId = applicationWithPrograms.getApplicationId();
+      List<PluginInstanceDetail> plugins = getPlugins(appId);
```

Comment:
getPlugins is getting the spec from the store. Since the spec is already available in ApplicationWithPrograms, that can be directly used to get the plugins.